simpleclient/src/main/java/io/prometheus/client/Summary.java

```diff
@@ -93,11 +124,80 @@ public double observeDuration() {
     public static class Value {
       public final double count;
       public final double sum;
+      public final double[] values;
 
-      private Value(double count, double sum) {
+      private Value(double count, double sum, double[] values) {
         this.count = count;
         this.sum = sum;
+        this.values = values;
       }
+
+      /**
+       * Return the φ-quantile estimation of the observed events.
+       * The approximate value is calculated by interpolation
+       * between adjacent elements in the array of observed values.
+       *
+       * @param φ a given quantile, in {@code [0, 1]}
+       * @return the approximate value of the the φ-quantile
+       */
+      public double getQuantile(double φ) {
+        if (φ < 0 || φ > 1) {
+          throw new IllegalArgumentException("quantile value must be in interval [0, 1]");
+        }
+
+        final int length = values.length;
+
+        if (length == 0) {
+          return 0;
+        }
+
+        if (length == 1) {
+          return values[0];
+        }
+
+        final double index = index(φ);
+
+        if (index < 1) {
+          return values[0];
+        }
+
+        if (index >= length) {
+          return values[length - 1];
+        }
+
+        return approxQuantile(index);
+      }
+
+      /**
+       * Find the index of the element of the array
+       * that can be used to compute the φ-quantile value.
+       *
+       * @param φ a given quantile
+       * @return the index for the array of values
+       */
+      private double index(final double φ) {
+        final int length = values.length;
+        return Double.compare(φ, 0) == 0 ? 0 :
+               Double.compare(φ, 1) == 0 ? length :
+               φ * (length + 1);
+      }
+
+      /**
+       * Calculate an approximate quantile from the observed
+       * value at {@code index}.
+       *
+       * @param index the index in the array of observed values
+       * @return the approximate quantile value
+       */
+      private double approxQuantile(final double index) {
```

Comment:
Splitting this code makes it a bit tricky to verify that the preconditions are satisfied.